tracing.go

```
@@ -84,53 +87,155 @@ func (s Span) ParentID() uint64 { return s.parentID }
-		childSpan.spanID = uint64(traceRng.Int63())
+}
+
+// startOutboundSpan creates a new tracing span to represent the outbound RPC call.
+// If the context already contains a span, it will be used as a parent, otherwise
+// a new root span is created.
+//
+// If the tracer supports Zipkin-style trace IDs, then call.callReq.Tracing is
+// initialized with those IDs. Otherwise it is assigned random values.
+func (c *Connection) startOutboundSpan(ctx context.Context, serviceName, methodName string, call *OutboundCall, startTime time.Time) opentracing.Span {
+	var parent opentracing.SpanContext // ok to be nil
+	if s := opentracing.SpanFromContext(ctx); s != nil {
+		parent = s.Context()
 	}
-	return childSpan
+	span := c.Tracer().StartSpan(
+		methodName,
+		opentracing.ChildOf(parent),
+		opentracing.StartTime(startTime),
+	)
+	if isTracingDisabled(ctx) {
+		ext.SamplingPriority.Set(span, 0)
+	}
+	ext.SpanKindRPCClient.Set(span)
+	ext.PeerService.Set(span, serviceName)
+	setPeerHostPort(span, c.remotePeerInfo.HostPort)
+	span.SetTag("as", call.callReq.Headers[ArgScheme])
+	call.callReq.Tracing.initFromOpenTracing(span)
+	return span
 }
 
-func (s *Span) sampleRootSpan(sampleRate float64) {
-	// Sampling only affects root spans
-	if s.ParentID() != 0 {
-		return
+// InjectOutboundSpan retrieves OpenTracing Span from `response`, where it is stored
+// when the outbound call is initiated. The tracing API is used to serialize the span
+// into the application `headers`, which will propagate tracing context to the server.
+// Returns modified headers containing serialized tracing context.
+func InjectOutboundSpan(response *OutboundCallResponse, headers map[string]string) map[string]string {
+	span := response.span
+	if span == nil {
+		return headers
+	}
+	if headers == nil {
+		headers = make(map[string]string)
+	}
+	carrier := opentracing.TextMapCarrier(headers)
+	if err := span.Tracer().Inject(span.Context(), opentracing.TextMap, carrier); err == nil {
+		return headers
```

this is the same as if there was an error. should we ignore the error using `_ = [...]` and put a comment for why we're ignoring it.